Continuous Failure: py37_native.test.unit._contextvars_propagation_test.ContextVarsPropagationTest

```
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.7/unittest/case.py", line 615, in run
    testMethod()
  File "/var/local/git/grpc/src/python/grpcio_tests/tests/unit/_contextvars_propagation_test.py", line 112, in test_propagation_to_auth_plugin
    response = stub(_REQUEST, wait_for_ready=True)
  File "/var/local/git/grpc/py37_native/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/var/local/git/grpc/py37_native/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Getting metadata from plugin failed with error: None != 24601"
	debug_error_string = "{"created":"@1621521604.321564241","description":"Getting metadata from plugin failed with error: None != 24601","file":"src/core/lib/security/credentials/plugin/plugin_credentials.cc","file_line":93,"grpc_status":14}"
>

[1mstdout:[0m

[1mstderr:[0m
ERROR:grpc._plugin_wrapping:AuthMetadataPluginCallback "<unit._contextvars_propagation_test.TestCallCredentials object at 0x7f207876feb8>" raised exception!
Traceback (most recent call last):
  File "/var/local/git/grpc/py37_native/lib/python3.7/site-packages/grpc/_plugin_wrapping.py", line 78, in __call__
    context, _AuthMetadataPluginCallback(callback_state, callback))
  File "/var/local/git/grpc/src/python/grpcio_tests/tests/unit/_contextvars_propagation_test.py", line 78, in __call__
    _EXPECTED_VALUE))
AssertionError: None != 24601
```

https://source.cloud.google.com/results/invocations/7bf104e3-7d15-4e26-b5fa-18b68f4266a1/targets/github%2Fgrpc%2Frun_tests%2Fpython_linux_opt_native%2Fpy37_native.test.unit._contextvars_propagation_test.ContextVarsPropagationTest/tests

--- a/src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi
@@ -58,7 +58,10 @@ cdef int _get_metadata(void *state,
         cb(user_data, NULL, 0, status, c_error_details)
   args = context.service_url, context.method_name, callback,
   plugin = <object>state
-  plugin._stored_ctx.run(_spawn_callback_async, plugin, args)
+  if plugin._stored_ctx is not None:
+    plugin._stored_ctx.run(_spawn_callback_async, plugin, args)
+  else:
+    _spawn_callback_async(<object>state, args)
   return 0  # Asynchronous return
 
 

--- a/src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi
@@ -57,7 +57,8 @@ cdef int _get_metadata(void *state,
       with nogil:
         cb(user_data, NULL, 0, status, c_error_details)
   args = context.service_url, context.method_name, callback,
-  _spawn_callback_async(<object>state, args)
+  plugin = <object>state
+  plugin._stored_ctx.run(_spawn_callback_async, plugin, args)
   return 0  # Asynchronous return
 
 

--- a/src/python/grpcio/grpc/_plugin_wrapping.py
+++ b/src/python/grpcio/grpc/_plugin_wrapping.py
@@ -68,6 +68,7 @@ class _Plugin(object):
 
     def __init__(self, metadata_plugin):
         self._metadata_plugin = metadata_plugin
+        self._stored_ctx = None
 
         try:
             import contextvars  # pylint: disable=wrong-import-position
@@ -77,7 +78,7 @@ class _Plugin(object):
             self._stored_ctx = contextvars.copy_context()
         except ImportError:
             # Support versions predating contextvars.
-            self._stored_ctx = None
+            pass
 
     def __call__(self, service_url, method_name, callback):
         context = _AuthMetadataContext(_common.decode(service_url),

--- a/src/python/grpcio/grpc/_plugin_wrapping.py
+++ b/src/python/grpcio/grpc/_plugin_wrapping.py
@@ -13,7 +13,6 @@
 # limitations under the License.
 
 import collections
-import contextvars
 import logging
 import threading
 
@@ -70,10 +69,15 @@ class _Plugin(object):
     def __init__(self, metadata_plugin):
         self._metadata_plugin = metadata_plugin
 
-        # The plugin may be invoked on a thread created by Core, which will not
-        # have the context propagated. This context is stored and installed in
-        # the thread invoking the plugin.
-        self._stored_ctx = contextvars.copy_context()
+        try:
+            import contextvars  # pylint: disable=wrong-import-position
+            # The plugin may be invoked on a thread created by Core, which will not
+            # have the context propagated. This context is stored and installed in
+            # the thread invoking the plugin.
+            self._stored_ctx = contextvars.copy_context()
+        except ImportError:
+            # Support versions predating contextvars.
+            self._stored_ctx = None
 
     def __call__(self, service_url, method_name, callback):
         context = _AuthMetadataContext(_common.decode(service_url),

--- a/src/python/grpcio/grpc/_plugin_wrapping.py
+++ b/src/python/grpcio/grpc/_plugin_wrapping.py
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 import collections
+import contextvars
 import logging
 import threading
 
@@ -68,6 +69,11 @@ class _Plugin(object):
 
     def __init__(self, metadata_plugin):
         self._metadata_plugin = metadata_plugin
+
+        # The plugin may be invoked on a thread created by Core, which will not
+        # have the context propagated. This context is stored and installed in
+        # the thread invoking the plugin.
+        self._stored_ctx = contextvars.copy_context()
 
     def __call__(self, service_url, method_name, callback):
         context = _AuthMetadataContext(_common.decode(service_url),

